packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -1,94 +1,107 @@
-import { DEFAULT_PLOT_CONFIG, pluginLog } from './utils';
+import { DEFAULT_PLOT_CONFIG } from './utils';
 import { PlotProps } from './types';
-import usePrevious from 'react-use/lib/usePrevious';
+
+function sameDims(prevProps: PlotProps, nextProps: PlotProps) {
+  return nextProps.width === prevProps.width && nextProps.height === prevProps.height;
+}
+
+function sameData(prevProps: PlotProps, nextProps: PlotProps) {
+  return nextProps.data === prevProps.data;
+}
+
+function sameConfig(prevProps: PlotProps, nextProps: PlotProps) {
+  return nextProps.config === prevProps.config;
+}
+
+type UPlotChartState = { ctx: { plot: uPlot | null } };
 
 /**
  * @internal
  * uPlot abstraction responsible for plot initialisation, setup and refresh
  * Receives a data frame that is x-axis aligned, as of https://github.com/leeoniya/uPlot/tree/master/docs#data-format
  * Exposes contexts for plugins registration and uPlot instance access
  */
-export const UPlotChart: React.FC<PlotProps> = (props) => {
-  const plotContainer = useRef<HTMLDivElement>(null);
-  const plotInstance = useRef<uPlot>();
-  const prevProps = usePrevious(props);
+export class UPlotChart extends React.Component<PlotProps, UPlotChartState> {
+  plotContainer = createRef<HTMLDivElement>();
 
-  const config = useMemo(() => {
-    return {
-      ...DEFAULT_PLOT_CONFIG,
-      width: props.width,
-      height: props.height,
-      ms: 1,
-      ...props.config.getConfig(),
-    } as uPlot.Options;
-  }, [props.config]);
-
-  useLayoutEffect(() => {
-    if (!plotInstance.current || props.width === 0 || props.height === 0) {
+  constructor(props: PlotProps) {
+    super(props);
+
+    this.state = {
+      ctx: {
+        plot: null,
+      },
+    };
+  }
+
+  reinitPlot() {
+    let { ctx } = this.state;
+
+    ctx.plot?.destroy();
+
+    let { width, height } = this.props;
+
+    if (width === 0 && height === 0) {
       return;
     }
 
-    pluginLog('uPlot core', false, 'updating size');
-    plotInstance.current.setSize({
-      width: props.width,
-      height: props.height,
+    const config: Options = {
+      ...DEFAULT_PLOT_CONFIG,
+      width: this.props.width,
+      height: this.props.height,
+      ms: 1 as 1,
+      ...this.props.config.getConfig(),
+    };
+
+    // for some reason does not invoke a re-render when called from componentDidMount or componentDidUpdate
+    this.setState({
+      ctx: {
+        plot: new uPlot(config, this.props.data, this.plotContainer!.current!),
+      },
     });
-  }, [props.width, props.height]);
 
-  // Effect responsible for uPlot updates/initialization logic. It's performed whenever component's props have changed
-  useLayoutEffect(() => {
-    // 0. Exit early if the component is not ready to initialize uPlot
-    if (!plotContainer.current || props.width === 0 || props.height === 0) {
-      return;
-    }
+    // force a re-render to update plugins' PlotContext with new instance
+    this.forceUpdate();
```

Comment:
You are setting the state few lines above, which already triggers the render so the context should be up to date - no need for forceUpdate

Reply:
it didnt work without this for some reason. there was no rerender and the context was stale in the children without this. :shrug:

Reply:
I've tested it and haven't noticed any issues. If you have steps to repro that would be awesome. Anyways - if there is such a problem, then I'm all for getters rather than properties to avoid the `forceUpdate`.  At least they always refer to the latest instance. Reading this code with setState and forceUpdate right after it is 🤯

Reply:
i found the issue that i was not diffing prev and next context in sCU(), which was preventing a re-render and a context refresh downstream. `forceUpdate()` is no more.